In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings("ignore")


# Load cleaned dataset
df = pd.read_csv(
    "../../data/processed/breast_cancer_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 569
Columns : 31


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================


target_column = "target"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]



print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")

display(y.head())

FEATURE MATRIX (X)
Rows    : 569
Columns : 30

TARGET VARIABLE (y)
Rows : 569


0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================


print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True)*100)
    .round(2)
)

TARGET DISTRIBUTION


target
1    357
0    212
Name: count, dtype: int64


Class Percentage


target
1    62.74
0    37.26
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================


print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
mean radius,float64
mean texture,float64
mean perimeter,float64
mean area,float64
mean smoothness,float64
mean compactness,float64
mean concavity,float64
mean concave points,float64
mean symmetry,float64
mean fractal dimension,float64


In [5]:
# =============================================================================
# STEP 3: TRAIN-TEST SPLIT
# =============================================================================


from sklearn.model_selection import train_test_split


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (455, 30)
y_train shape : (455,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (114, 30)
y_test shape  : (114,)


In [6]:
# =============================================================================
# CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================


print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True)*100)
    .round(2)
)



print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True)*100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


target
1    62.64
0    37.36
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


target
1    63.16
0    36.84
Name: proportion, dtype: float64

In [7]:
# =============================================================================
# STEP 4: FEATURE SCALING
# =============================================================================


from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


# Initialize scaler
scaler = StandardScaler()


# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)


# Transform test data using same scaler
X_test_scaled = scaler.transform(X_test)



print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(f"X_train_scaled shape : {X_train_scaled.shape}")
print(f"X_test_scaled shape  : {X_test_scaled.shape}")

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (455, 30)
X_test_scaled shape  : (114, 30)


In [8]:
# =============================================================================
# CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns
)



print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(X_train_scaled.head())

SCALED TRAIN DATA PREVIEW


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
0,-1.072001,-0.658425,-1.088080,-0.939274,-0.135940,-1.008718,-0.968359,-1.102032,0.281062,-0.113231,...,-1.034094,-0.623497,-1.070773,-0.876534,-0.169982,-1.038836,-1.078995,-1.350527,-0.352658,-0.541380
1,1.748743,0.066502,1.751157,1.745559,1.274468,0.842288,1.519852,1.994664,-0.293045,-0.320180,...,1.228342,-0.092833,1.187467,1.104386,1.517001,0.249655,1.178594,1.549916,0.191078,-0.173739
2,-0.974734,-0.931124,-0.997709,-0.867589,-0.613515,-1.138154,-1.092292,-1.243358,0.434395,-0.429247,...,-0.973231,-1.036772,-1.008044,-0.834168,-1.097823,-1.167260,-1.282241,-1.707442,-0.307734,-1.213033
3,-0.145103,-1.215186,-0.123013,-0.253192,0.664482,0.286762,-0.129729,-0.098605,0.555635,0.029395,...,-0.251266,-1.369643,-0.166633,-0.330292,0.234006,0.096874,-0.087521,-0.344838,0.242198,-0.118266
4,-0.771617,-0.081211,-0.803700,-0.732927,-0.672282,-1.006099,-0.798502,-0.684484,0.737495,-0.457213,...,-0.801135,0.079230,-0.824381,-0.741830,-0.911367,-0.984612,-0.933190,-0.777604,0.555118,-0.761639


In [9]:
# =============================================================================
# SCALING VERIFICATION
# =============================================================================


print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.mean().round(2)
)



print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


mean radius               -0.0
mean texture               0.0
mean perimeter            -0.0
mean area                 -0.0
mean smoothness            0.0
mean compactness          -0.0
mean concavity            -0.0
mean concave points        0.0
mean symmetry             -0.0
mean fractal dimension     0.0
radius error              -0.0
texture error             -0.0
perimeter error            0.0
area error                 0.0
smoothness error          -0.0
compactness error         -0.0
concavity error            0.0
concave points error      -0.0
symmetry error             0.0
fractal dimension error    0.0
worst radius               0.0
worst texture             -0.0
worst perimeter           -0.0
worst area                 0.0
worst smoothness           0.0
worst compactness         -0.0
worst concavity            0.0
worst concave points      -0.0
worst symmetry            -0.0
worst fractal dimension    0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


mean radius                1.0
mean texture               1.0
mean perimeter             1.0
mean area                  1.0
mean smoothness            1.0
mean compactness           1.0
mean concavity             1.0
mean concave points        1.0
mean symmetry              1.0
mean fractal dimension     1.0
radius error               1.0
texture error              1.0
perimeter error            1.0
area error                 1.0
smoothness error           1.0
compactness error          1.0
concavity error            1.0
concave points error       1.0
symmetry error             1.0
fractal dimension error    1.0
worst radius               1.0
worst texture              1.0
worst perimeter            1.0
worst area                 1.0
worst smoothness           1.0
worst compactness          1.0
worst concavity            1.0
worst concave points       1.0
worst symmetry             1.0
worst fractal dimension    1.0
dtype: float64

# LASSO (L1-regularized Logistic Regression)

## 2. Baseline

In [10]:
# =============================================================================
# LASSO BASELINE FEATURE SELECTION (TOP-K = 5,10,15,20)
# =============================================================================

from sklearn.linear_model import LogisticRegression
import pandas as pd
import numpy as np


# Fit LASSO once with all features
# Logistic Regression with L1 penalty = LASSO-style feature selection

lasso_selector = LogisticRegression(
    penalty="l1",
    solver="liblinear",
    C=0.1,
    max_iter=2000,
    class_weight="balanced",
    random_state=42
)


lasso_selector.fit(
    X_train_scaled,
    y_train
)


# =============================================================================
# Extract LASSO coefficients
# =============================================================================

lasso_coefficients = lasso_selector.coef_[0]


# Use absolute coefficient magnitude as feature importance
lasso_scores = pd.DataFrame({

    "Feature": X_train_scaled.columns,

    "LASSO Coefficient": lasso_coefficients,

    "LASSO Score": np.abs(lasso_coefficients)

})


# Sort by absolute coefficient magnitude
lasso_scores = lasso_scores.sort_values(
    by="LASSO Score",
    ascending=False
).reset_index(drop=True)


# =============================================================================
# Top-K values
# =============================================================================

k_values = [5, 10, 15, 20]

lasso_results = {}


for k in k_values:

    top_features = (
        lasso_scores
        .head(k)["Feature"]
        .tolist()
    )


    # Boolean mask for selecting columns
    mask = X_train_scaled.columns.isin(
        top_features
    )


    X_train_selected = X_train_scaled.loc[
        :, mask
    ]

    X_test_selected = X_test_scaled.loc[
        :, mask
    ]


    lasso_results[k] = {

        "features": top_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "scores": lasso_scores.head(k)

    }


    print("=" * 70)
    print(f"TOP {k} LASSO FEATURES")
    print("=" * 70)

    print(top_features)

    print("\nScores:")
    print(lasso_scores.head(k))

TOP 5 LASSO FEATURES
['worst radius', 'worst concave points', 'worst texture', 'radius error', 'mean concave points']

Scores:
                Feature  LASSO Coefficient  LASSO Score
0          worst radius          -2.079876     2.079876
1  worst concave points          -1.084678     1.084678
2         worst texture          -0.626262     0.626262
3          radius error          -0.517377     0.517377
4   mean concave points          -0.405840     0.405840
TOP 10 LASSO FEATURES
['worst radius', 'worst concave points', 'worst texture', 'radius error', 'mean concave points', 'worst symmetry', 'worst smoothness', 'mean texture', 'mean symmetry', 'mean smoothness']

Scores:
                Feature  LASSO Coefficient  LASSO Score
0          worst radius          -2.079876     2.079876
1  worst concave points          -1.084678     1.084678
2         worst texture          -0.626262     0.626262
3          radius error          -0.517377     0.517377
4   mean concave points          -0.405

In [11]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {

    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,
        eval_metric="logloss",
        scale_pos_weight=(
            (y_train == 0).sum() /
            (y_train == 1).sum()
        )
    )
}

In [12]:
# =============================================================================
# MODEL EVALUATION — LASSO
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

results_lasso = []


for k in k_values:

    Xtr = lasso_results[k]["X_train"]

    Xts = lasso_results[k]["X_test"]

    features = ", ".join(
        lasso_results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        results_lasso.append({

            "Method": "LASSO",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred
            ),

            "Recall": recall_score(
                y_test,
                y_pred
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )

        })


lasso_evaluation = pd.DataFrame(
    results_lasso
)

display(lasso_evaluation)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,LASSO,5,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.993386
1,LASSO,5,Random Forest,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.995040
2,LASSO,5,XGBoost,"worst radius, worst concave points, worst text...",0.947368,0.958333,0.958333,0.958333,0.994048
3,LASSO,10,Logistic Regression,"worst radius, worst concave points, worst text...",0.973684,0.985915,0.972222,0.979021,0.996032
4,LASSO,10,Random Forest,"worst radius, worst concave points, worst text...",0.964912,0.985714,0.958333,0.971831,0.994709
5,LASSO,10,XGBoost,"worst radius, worst concave points, worst text...",0.956140,0.958904,0.972222,0.965517,0.992394
6,LASSO,15,Logistic Regression,"worst radius, worst concave points, worst text...",0.947368,0.985294,0.930556,0.957143,0.993717
7,LASSO,15,Random Forest,"worst radius, worst concave points, worst text...",0.947368,0.971429,0.944444,0.957746,0.994709
8,LASSO,15,XGBoost,"worst radius, worst concave points, worst text...",0.964912,0.972222,0.972222,0.972222,0.995040
9,LASSO,20,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.985507,0.944444,0.964539,0.994709


In [13]:
lasso_evaluation.to_csv(
    "../../../results/breast_cancer/baseline_results/lasso_baseline_results.csv",
    index=False
)

## 2. DODA

In [14]:
%pip uninstall -y doda

Found existing installation: doda 0.1.0
Uninstalling doda-0.1.0:
  Successfully uninstalled doda-0.1.0
Note: you may need to restart the kernel to use updated packages.


In [15]:
%pip install git+https://github.com/anandha-3679/DODA.git

  Cloning https://github.com/anandha-3679/DODA.git to C:\Users\johnm\AppData\Local\Temp\pip-req-build-cuhsa8n6
  Resolved https://github.com/anandha-3679/DODA.git to commit bb479647305e6c7166c6d8a0c4f93d6291f36998
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
  Created wheel for doda: filename=doda-0.1.0-py3-none-any.whl size=14746 sha256=d38ea6042d382b0539a52c912258fa037accab01c40a00e2ada366119b06d1a9
  Stored in directory: C:\Users\johnm\AppData\Local\Temp\pip-ephem-wheel-cache-8_9lh53t\wheels\42\6a\22\a51a03690cc18d3493705062872e3519eede94dc00682c19bc
Successfully built doda
Note: you may need to restart the kernel to use updated packages.


  Running command git clone --filter=blob:none --quiet https://github.com/anandha-3679/DODA.git 'C:\Users\johnm\AppData\Local\Temp\pip-req-build-cuhsa8n6'


In [16]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [17]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\johnm\msc_research\.venv\Scripts\python.exe
DODA: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\__init__.py
Selector: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\selector.py


In [18]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import HadamardFusion

In [19]:
# =============================================================================
# LASSO + DODA FEATURE SELECTION
# =============================================================================

import pandas as pd
from sklearn.linear_model import LogisticRegression


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "../../../config/clinical_weights/breast_cancer.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = HadamardFusion()


# -------------------------------------------------------------------------
# Store DODA results
# -------------------------------------------------------------------------

doda_results = {}


# =============================================================================
# RUN LASSO + DODA FOR EACH TOP-K
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"LASSO + DODA \u2014 TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # LASSO operator
    # ---------------------------------------------------------------------

    lasso = SklearnAdapter(

        LogisticRegression(
            penalty="l1",
            solver="liblinear",
            C=0.1,
            max_iter=2000,
            class_weight="balanced",
            random_state=42
        )

    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(

        operators=[lasso],

        provider=provider,

        fusion=fusion,

        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on TRAINING data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(

        X_train_scaled,

        y_train

    )


    # ---------------------------------------------------------------------
    # Transform TEST data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(

        X_test_scaled

    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = (
        selector.get_selected_features()
    )


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store everything needed for later evaluation
    # ---------------------------------------------------------------------

    doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores":
            selector.raw_math_scores_,

        "math_scores":
            selector.math_scores_,

        "clinical_weights":
            selector.clinical_weights_,

        "final_scores":
            selector.final_scores_
    }


print("\n" + "=" * 70)
print("LASSO + DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


LASSO + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x000001AC4C475160>
DODA Pipeline Started
Running: LogisticRegression
DODA feature order:
['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension']
Number of features: 30
Number of coefficients: 30
Operator Scores:
{'LogisticRegression_1': {'mean radius': np.float64(0.0), 'mean texture': np.float64(0.03116243353789412), 'mean perimeter': np.float64(0.0), 'mean area': np.float64(0.0), 'mean smoothness': np.float64(0.0), 'mean compactness': np.float64(0.0), 'mean concavity': np.float64(0.0), 'mean concave points': np.float64(0.4058401456180695), 'mean symmetry': np.float64(0.0), 'mean fractal dimension': np.float64(0.0), 'radius error': np.float64(0.5173770695234363), 'texture error': np.float64(0.0), 'perimeter error': np.float64(0.0),

In [20]:
# CELL 2
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = doda_results[k]["selector"]

# -----------------------------------------------------------------------------
# Raw mathematical scores
# -----------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Normalized mathematical scores
# -----------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Combine both
# -----------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)

math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)

print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
20,worst radius,2.079876,1.000000
27,worst concave points,1.084678,0.521511
21,worst texture,0.626262,0.301106
10,radius error,0.517377,0.248754
7,mean concave points,0.405840,0.195127
28,worst symmetry,0.152786,0.073459
24,worst smoothness,0.151605,0.072891
1,mean texture,0.031162,0.014983
8,mean symmetry,0.000000,0.000000
4,mean smoothness,0.000000,0.000000


In [21]:
# CELL 3
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)

clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)

print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
27,worst concave points,1.0
26,worst concavity,1.0
22,worst perimeter,1.0
23,worst area,1.0
20,worst radius,1.0
0,mean radius,0.9
3,mean area,0.9
6,mean concavity,0.9
7,mean concave points,0.9
2,mean perimeter,0.9


In [22]:
# CELL 4
# =============================================================================
# INSPECT FINAL DODA SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final DODA Score"
    ]
)

final_scores = final_scores.sort_values(
    by="Final DODA Score",
    ascending=False
)

print("=" * 70)
print("FINAL DODA SCORES")
print("=" * 70)

display(final_scores)

FINAL DODA SCORES


,Feature,Final DODA Score
20,worst radius,1.000000
27,worst concave points,0.521511
21,worst texture,0.180663
7,mean concave points,0.175614
10,radius error,0.124377
28,worst symmetry,0.044075
24,worst smoothness,0.036446
1,mean texture,0.007491
8,mean symmetry,0.000000
4,mean smoothness,0.000000


In [23]:
selector.get_selected_features()

['worst radius',
 'worst concave points',
 'worst texture',
 'mean concave points',
 'radius error',
 'worst symmetry',
 'worst smoothness',
 'mean texture',
 'mean radius',
 'mean perimeter']

In [24]:
comparison = selector.compare_scores()

comparison


MATHEMATICAL vs CLINICAL vs HADAMARDFUSION

Fusion Method: HadamardFusion

Top Features After Fusion:
                Feature  Math Rank  Clinical Rank  Final Rank  Rank Change
0          worst radius          1              1           1            0
1  worst concave points          2              5           2            0
2         worst texture          3             13           3            0
3   mean concave points          5             10           4            1
4          radius error          4             17           5           -1
5        worst symmetry          6             14           6            0
6      worst smoothness          7             20           7            0
7          mean texture          8             15           8            0
8           mean radius          9              6           9            0
9        mean perimeter         10              7          10            0


,Feature,Math Rank,Normalized Math Score,Clinical Rank,Clinical Weight,Final Rank,Final Score,Rank Change
0,worst radius,1,1.0000,1,1.0,1,1.0000,0
1,worst concave points,2,0.5215,5,1.0,2,0.5215,0
2,worst texture,3,0.3011,13,0.6,3,0.1807,0
3,mean concave points,5,0.1951,10,0.9,4,0.1756,1
4,radius error,4,0.2488,17,0.5,5,0.1244,-1
5,worst symmetry,6,0.0735,14,0.6,6,0.0441,0
6,worst smoothness,7,0.0729,20,0.5,7,0.0364,0
7,mean texture,8,0.0150,15,0.5,8,0.0075,0
8,mean radius,9,0.0000,6,0.9,9,0.0000,0
9,mean perimeter,10,0.0000,7,0.9,10,0.0000,0


In [25]:
# =============================================================================
# LASSO + DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


# -------------------------------------------------------------------------
# Store results
# -------------------------------------------------------------------------

lasso_doda_model_results = []


# -------------------------------------------------------------------------
# Evaluate for each Top-K
# -------------------------------------------------------------------------

for k in k_values:

    Xtr = doda_results[k]["X_train"]

    Xts = doda_results[k]["X_test"]

    features = ", ".join(
        doda_results[k]["features"]
    )


    # ---------------------------------------------------------------------
    # Evaluate each model
    # ---------------------------------------------------------------------

    for model_name, model in models.items():

        # Train model
        model.fit(
            Xtr,
            y_train
        )


        # Predictions
        y_pred = model.predict(
            Xts
        )


        # Probability predictions
        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        # ---------------------------------------------------------------
        # Store evaluation metrics
        # ---------------------------------------------------------------

        lasso_doda_model_results.append({

            "Method": "LASSO + DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


# -------------------------------------------------------------------------
# Convert results to DataFrame
# -------------------------------------------------------------------------

lasso_doda_results_df = pd.DataFrame(
    lasso_doda_model_results
)


# -------------------------------------------------------------------------
# Display results
# -------------------------------------------------------------------------

display(
    lasso_doda_results_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,LASSO + DODA,5,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.993386
1,LASSO + DODA,5,Random Forest,"worst radius, worst concave points, worst text...",0.956140,0.958904,0.972222,0.965517,0.995040
2,LASSO + DODA,5,XGBoost,"worst radius, worst concave points, worst text...",0.964912,0.972222,0.972222,0.972222,0.993717
3,LASSO + DODA,10,Logistic Regression,"worst radius, worst concave points, worst text...",0.938596,0.985075,0.916667,0.949640,0.994709
4,LASSO + DODA,10,Random Forest,"worst radius, worst concave points, worst text...",0.938596,0.971014,0.930556,0.950355,0.994709
5,LASSO + DODA,10,XGBoost,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.994709
6,LASSO + DODA,15,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.985507,0.944444,0.964539,0.993717
7,LASSO + DODA,15,Random Forest,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.994213
8,LASSO + DODA,15,XGBoost,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.994378
9,LASSO + DODA,20,Logistic Regression,"worst radius, worst concave points, worst text...",0.947368,0.985294,0.930556,0.957143,0.994709


In [26]:
lasso_doda_results_df.to_csv(
    "../../../results/breast_cancer/doda_results/lasso_doda_results.csv",
    index=False
)

In [27]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================

comparison_df = pd.concat(
    [
        lasso_evaluation,
        lasso_doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,LASSO,5,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.993386
1,LASSO,5,Random Forest,"worst radius, worst concave points, worst text...",0.956140,0.971831,0.958333,0.965035,0.995040
2,LASSO,5,XGBoost,"worst radius, worst concave points, worst text...",0.947368,0.958333,0.958333,0.958333,0.994048
3,LASSO,10,Logistic Regression,"worst radius, worst concave points, worst text...",0.973684,0.985915,0.972222,0.979021,0.996032
4,LASSO,10,Random Forest,"worst radius, worst concave points, worst text...",0.964912,0.985714,0.958333,0.971831,0.994709
5,LASSO,10,XGBoost,"worst radius, worst concave points, worst text...",0.956140,0.958904,0.972222,0.965517,0.992394
6,LASSO,15,Logistic Regression,"worst radius, worst concave points, worst text...",0.947368,0.985294,0.930556,0.957143,0.993717
7,LASSO,15,Random Forest,"worst radius, worst concave points, worst text...",0.947368,0.971429,0.944444,0.957746,0.994709
8,LASSO,15,XGBoost,"worst radius, worst concave points, worst text...",0.964912,0.972222,0.972222,0.972222,0.995040
9,LASSO,20,Logistic Regression,"worst radius, worst concave points, worst text...",0.956140,0.985507,0.944444,0.964539,0.994709


In [28]:
for k in k_values:

    selector = doda_results[k]["selector"]

    raw_rank = [
        x[0]
        for x in sorted(
            selector.raw_math_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    doda_rank = [
        x[0]
        for x in sorted(
            selector.final_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    print("\n" + "=" * 70)
    print(f"TOP {k}")
    print("=" * 70)

    print("Raw Math Ranking:")
    print(raw_rank)

    print("\nDODA Ranking:")
    print(doda_rank)

    print("\nTop-K Same:",
          raw_rank[:k] == doda_rank[:k])


TOP 5
Raw Math Ranking:
['worst radius', 'worst concave points', 'worst texture', 'radius error', 'mean concave points', 'worst symmetry', 'worst smoothness', 'mean texture', 'mean radius', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean symmetry', 'mean fractal dimension', 'texture error', 'perimeter error', 'area error', 'smoothness error', 'compactness error', 'concavity error', 'concave points error', 'symmetry error', 'fractal dimension error', 'worst perimeter', 'worst area', 'worst compactness', 'worst concavity', 'worst fractal dimension']

DODA Ranking:
['worst radius', 'worst concave points', 'worst texture', 'mean concave points', 'radius error', 'worst symmetry', 'worst smoothness', 'mean texture', 'mean radius', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean symmetry', 'mean fractal dimension', 'texture error', 'perimeter error', 'area error', 'smoothness error', 'compactness erro

In [29]:
# =============================================================================
# STEP 21: CROSS-VALIDATION CONFIGURATION
# =============================================================================
# LASSO vs LASSO + DODA
#
# The fixed-split experiment above is retained unchanged.
# This section adds fold-wise and repeated-CV validation.
#
# LASSO requires standardized predictors. Therefore, scaling is fitted
# separately inside every CV training fold to prevent data leakage.
#
# DODA continues to use the existing Hadamard fusion method exactly as
# defined in the fixed-split experiment.

import os
import numpy as np
import pandas as pd

from itertools import combinations

from sklearn.model_selection import (
    StratifiedKFold,
    RepeatedStratifiedKFold
)

from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

from doda import DODASelector
from doda.adapters import SklearnAdapter
from doda.knowledge.providers import JSONProvider
from doda.fusion import HadamardFusion


# -------------------------------------------------------------------------
# Research design
# -------------------------------------------------------------------------

TOP_K_VALUES = [5, 10, 15, 20]

N_SPLITS = 5
N_REPEATS = 5
RANDOM_STATE = 42

WEIGHTS_FILE = "../../../config/clinical_weights/breast_cancer.json"

print("=" * 80)
print("BREAST CANCER LASSO CROSS-VALIDATION CONFIGURATION")
print("=" * 80)

print("Top-K values:", TOP_K_VALUES)
print("5-fold CV:", N_SPLITS)
print(
    "Repeated CV:",
    f"{N_SPLITS} folds × {N_REPEATS} repeats"
)
print(
    "Total repeated runs:",
    N_SPLITS * N_REPEATS
)
print("Random state:", RANDOM_STATE)
print("Clinical weights:", WEIGHTS_FILE)


# -------------------------------------------------------------------------
# Fresh LASSO factory
# -------------------------------------------------------------------------

def make_lasso(random_state=RANDOM_STATE):
    """Create a fresh L1-regularized logistic regression selector."""

    return LogisticRegression(
        penalty="l1",
        solver="liblinear",
        C=0.1,
        max_iter=2000,
        class_weight="balanced",
        random_state=random_state
    )


def make_doda_lasso(top_k):
    """Create a fresh LASSO + DODA selector using Hadamard fusion."""

    lasso_operator = SklearnAdapter(
        make_lasso(RANDOM_STATE)
    )

    provider = JSONProvider(
        WEIGHTS_FILE
    )

    return DODASelector(
        operators=[lasso_operator],
        provider=provider,
        fusion=HadamardFusion(),
        top_k=top_k
    )


# -------------------------------------------------------------------------
# Downstream models
# -------------------------------------------------------------------------

cv_models = {

    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=RANDOM_STATE
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=RANDOM_STATE,
        n_jobs=-1
    ),

    "XGBoost": XGBClassifier(
        random_state=RANDOM_STATE,
        eval_metric="logloss",
        scale_pos_weight=(
            (y == 0).sum() /
            (y == 1).sum()
        )
    )
}

print("\nCross-validation configuration loaded successfully.")


BREAST CANCER LASSO CROSS-VALIDATION CONFIGURATION
Top-K values: [5, 10, 15, 20]
5-fold CV: 5
Repeated CV: 5 folds × 5 repeats
Total repeated runs: 25
Random state: 42
Clinical weights: ../../../config/clinical_weights/breast_cancer.json

Cross-validation configuration loaded successfully.


In [30]:
# =============================================================================
# STEP 22: 5-FOLD CV PREDICTIVE PERFORMANCE
# LASSO vs LASSO + DODA
# =============================================================================
# Feature selection is fitted inside each training fold.
# Scaling is also fitted inside each training fold.
#
# This prevents information from the validation fold from influencing
# LASSO feature selection, DODA ranking, or downstream Logistic Regression.

cv = StratifiedKFold(
    n_splits=N_SPLITS,
    shuffle=True,
    random_state=RANDOM_STATE
)

cv_results = []


for method in [
    "LASSO",
    "LASSO + DODA"
]:

    print("\n" + "=" * 80)
    print(f"METHOD: {method}")
    print("=" * 80)

    for TOP_K in TOP_K_VALUES:

        for fold, (train_idx, val_idx) in enumerate(
            cv.split(X, y),
            start=1
        ):

            X_train_cv = X.iloc[train_idx].copy()
            X_val_cv = X.iloc[val_idx].copy()

            y_train_cv = y.iloc[train_idx].copy()
            y_val_cv = y.iloc[val_idx].copy()


            # -------------------------------------------------------------
            # SCALE WITH TRAINING-FOLD STATISTICS ONLY
            # -------------------------------------------------------------

            fold_scaler = StandardScaler()

            X_train_cv_scaled = pd.DataFrame(
                fold_scaler.fit_transform(X_train_cv),
                columns=X_train_cv.columns,
                index=X_train_cv.index
            )

            X_val_cv_scaled = pd.DataFrame(
                fold_scaler.transform(X_val_cv),
                columns=X_val_cv.columns,
                index=X_val_cv.index
            )


            # -------------------------------------------------------------
            # FEATURE SELECTION
            # -------------------------------------------------------------

            if method == "LASSO":

                lasso_selector_cv = make_lasso()

                lasso_selector_cv.fit(
                    X_train_cv_scaled,
                    y_train_cv
                )

                lasso_scores_cv = pd.DataFrame({
                    "Feature": X_train_cv.columns,
                    "LASSO_Coefficient":
                        lasso_selector_cv.coef_[0]
                })

                lasso_scores_cv["LASSO_Score"] = (
                    lasso_scores_cv[
                        "LASSO_Coefficient"
                    ].abs()
                )

                # Stable tie handling for reproducibility.
                lasso_scores_cv = (
                    lasso_scores_cv
                    .sort_values(
                        by=[
                            "LASSO_Score",
                            "Feature"
                        ],
                        ascending=[
                            False,
                            True
                        ]
                    )
                    .reset_index(drop=True)
                )

                selected_features = (
                    lasso_scores_cv
                    .head(TOP_K)["Feature"]
                    .tolist()
                )

            else:

                # DODA receives the complete scaled training matrix.
                # The LASSO operator supplies the statistical ranking,
                # while DODA applies the existing Hadamard fusion with
                # the clinical weights before selecting Top-K.

                doda_selector_cv = make_doda_lasso(
                    TOP_K
                )

                doda_selector_cv.fit(
                    X_train_cv_scaled,
                    y_train_cv
                )

                selected_features = list(
                    doda_selector_cv.get_selected_features()
                )


            # -------------------------------------------------------------
            # Selected train/validation matrices
            # -------------------------------------------------------------

            X_train_selected = X_train_cv_scaled[
                selected_features
            ]

            X_val_selected = X_val_cv_scaled[
                selected_features
            ]


            # -------------------------------------------------------------
            # Evaluate downstream models
            # -------------------------------------------------------------

            for model_name, model_template in cv_models.items():

                model = type(model_template)(
                    **model_template.get_params()
                )

                model.fit(
                    X_train_selected,
                    y_train_cv
                )

                y_pred = model.predict(
                    X_val_selected
                )

                y_prob = model.predict_proba(
                    X_val_selected
                )[:, 1]


                cv_results.append({

                    "Fold": fold,
                    "Method": method,
                    "Top_K": TOP_K,
                    "Model": model_name,

                    "Accuracy": accuracy_score(
                        y_val_cv,
                        y_pred
                    ),

                    "Precision": precision_score(
                        y_val_cv,
                        y_pred,
                        zero_division=0
                    ),

                    "Recall": recall_score(
                        y_val_cv,
                        y_pred,
                        zero_division=0
                    ),

                    "F1": f1_score(
                        y_val_cv,
                        y_pred,
                        zero_division=0
                    ),

                    "ROC_AUC": roc_auc_score(
                        y_val_cv,
                        y_prob
                    ),

                    "Selected_Features": ", ".join(
                        selected_features
                    )
                })


cv_results_df = pd.DataFrame(
    cv_results
)

print("\nCompleted 5-fold CV.")
print("Rows:", len(cv_results_df))

display(
    cv_results_df.head()
)



METHOD: LASSO

METHOD: LASSO + DODA
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x000001AC51529590>
DODA Pipeline Started
Running: LogisticRegression
DODA feature order:
['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension']
Number of features: 30
Number of coefficients: 30
Operator Scores:
{'LogisticRegression_1': {'mean radius': np.float64(0.0), 'mean texture': np.float64(0.0), 'mean perimeter': np.float64(0.0), 'mean area': np.float64(0.0), 'mean smoothness': np.float64(0.0), 'mean compactness': np.float64(0.0), 'mean concavity': np.float64(0.0), 'mean concave points': np.float64(0.3919984806350876), 'mean symmetry': np.float64(0.0), 'mean fractal dimension': np.float64(0.0), 'radius error': np.float64(0.5389807424230852), 'texture error': np.float64(0.0), 'perimeter error': np.float64(0.0), 'area err

,Fold,Method,Top_K,Model,Accuracy,Precision,Recall,F1,ROC_AUC,Selected_Features
0,1,LASSO,5,Logistic Regression,0.947368,0.985075,0.929577,0.956522,0.995742,"worst radius, worst concave points, worst text..."
1,1,LASSO,5,Random Forest,0.956140,0.985294,0.943662,0.964029,0.997052,"worst radius, worst concave points, worst text..."
2,1,LASSO,5,XGBoost,0.947368,0.971014,0.943662,0.957143,0.995414,"worst radius, worst concave points, worst text..."
3,2,LASSO,5,Logistic Regression,0.956140,0.945946,0.985915,0.965517,0.986571,"worst radius, worst concave points, worst text..."
4,2,LASSO,5,Random Forest,0.947368,0.945205,0.971831,0.958333,0.975762,"worst radius, worst concave points, worst text..."


In [31]:
# =============================================================================
# STEP 23: 5-FOLD CV SUMMARY — MEAN ± STD
# =============================================================================

cv_summary_df = (
    cv_results_df
    .groupby(
        ["Method", "Top_K", "Model"]
    )
    .agg(
        Accuracy_Mean=("Accuracy", "mean"),
        Accuracy_STD=("Accuracy", "std"),

        Precision_Mean=("Precision", "mean"),
        Precision_STD=("Precision", "std"),

        Recall_Mean=("Recall", "mean"),
        Recall_STD=("Recall", "std"),

        F1_Mean=("F1", "mean"),
        F1_STD=("F1", "std"),

        ROC_AUC_Mean=("ROC_AUC", "mean"),
        ROC_AUC_STD=("ROC_AUC", "std")
    )
    .reset_index()
)

print("=" * 100)
print("5-FOLD CV PERFORMANCE SUMMARY")
print("=" * 100)

display(
    cv_summary_df.round(4)
)

os.makedirs(
    "../../../results/breast_cancer/cv",
    exist_ok=True
)

cv_results_df.to_csv(
    "../../../results/breast_cancer/cv/lasso_doda_5fold_results.csv",
    index=False
)

cv_summary_df.to_csv(
    "../../../results/breast_cancer/cv/lasso_doda_5fold_summary.csv",
    index=False
)


5-FOLD CV PERFORMANCE SUMMARY


,Method,Top_K,Model,Accuracy_Mean,Accuracy_STD,Precision_Mean,Precision_STD,Recall_Mean,Recall_STD,F1_Mean,F1_STD,ROC_AUC_Mean,ROC_AUC_STD
0,LASSO,5,Logistic Regression,0.9578,0.0095,0.9724,0.0212,0.9608,0.0286,0.9661,0.0081,0.9939,0.0051
1,LASSO,5,Random Forest,0.9561,0.0106,0.9699,0.0240,0.9608,0.0301,0.9648,0.0089,0.9898,0.0094
2,LASSO,5,XGBoost,0.9526,0.0147,0.9695,0.0178,0.9553,0.0409,0.9616,0.0131,0.9925,0.0056
3,LASSO,10,Logistic Regression,0.9684,0.0100,0.9750,0.0151,0.9748,0.0117,0.9748,0.0081,0.9959,0.0043
4,LASSO,10,Random Forest,0.9579,0.0168,0.9806,0.0209,0.9524,0.0363,0.9657,0.0142,0.9936,0.0037
5,LASSO,10,XGBoost,0.9649,0.0062,0.9804,0.0124,0.9637,0.0210,0.9717,0.0053,0.9950,0.0041
6,LASSO,15,Logistic Regression,0.9702,0.0159,0.9775,0.0078,0.9748,0.0209,0.9761,0.0131,0.9954,0.0050
7,LASSO,15,Random Forest,0.9631,0.0190,0.9805,0.0154,0.9609,0.0386,0.9701,0.0160,0.9933,0.0041
8,LASSO,15,XGBoost,0.9684,0.0079,0.9805,0.0156,0.9692,0.0181,0.9746,0.0063,0.9958,0.0035
9,LASSO,20,Logistic Regression,0.9772,0.0159,0.9832,0.0115,0.9803,0.0236,0.9816,0.0131,0.9959,0.0055


In [32]:
# =============================================================================
# STEP 24: 5×5 REPEATED STRATIFIED CV
# LASSO vs LASSO + DODA
# =============================================================================
# 5 folds × 5 repeats = 25 train/validation partitions.
# LASSO and DODA are refit inside every repeated-CV partition.
#
# Scaling is also refit inside every partition using training data only.
#
# The repeated-CV results are stored separately from the fixed-split
# baseline and DODA result tables.

repeated_cv = RepeatedStratifiedKFold(
    n_splits=N_SPLITS,
    n_repeats=N_REPEATS,
    random_state=RANDOM_STATE
)

repeated_cv_results = []


for run_id, (train_idx, val_idx) in enumerate(
    repeated_cv.split(X, y),
    start=1
):

    print("=" * 70)
    print(
        f"RUN {run_id}/{N_SPLITS * N_REPEATS}"
    )
    print("=" * 70)

    X_train_rep = X.iloc[train_idx].copy()
    X_val_rep = X.iloc[val_idx].copy()

    y_train_rep = y.iloc[train_idx].copy()
    y_val_rep = y.iloc[val_idx].copy()


    # ---------------------------------------------------------------------
    # Fold-specific scaling
    # ---------------------------------------------------------------------

    rep_scaler = StandardScaler()

    X_train_rep_scaled = pd.DataFrame(
        rep_scaler.fit_transform(X_train_rep),
        columns=X_train_rep.columns,
        index=X_train_rep.index
    )

    X_val_rep_scaled = pd.DataFrame(
        rep_scaler.transform(X_val_rep),
        columns=X_val_rep.columns,
        index=X_val_rep.index
    )


    # ---------------------------------------------------------------------
    # Fit LASSO once per repeated-CV partition.
    # The same LASSO ranking is reused across Top-K values.
    # ---------------------------------------------------------------------

    lasso_selector_rep = make_lasso()

    lasso_selector_rep.fit(
        X_train_rep_scaled,
        y_train_rep
    )

    lasso_rank_rep = pd.DataFrame({
        "Feature": X_train_rep.columns,
        "LASSO_Coefficient":
            lasso_selector_rep.coef_[0]
    })

    lasso_rank_rep["LASSO_Score"] = (
        lasso_rank_rep[
            "LASSO_Coefficient"
        ].abs()
    )

    lasso_rank_rep = (
        lasso_rank_rep
        .sort_values(
            by=[
                "LASSO_Score",
                "Feature"
            ],
            ascending=[
                False,
                True
            ]
        )
        .reset_index(drop=True)
    )


    for TOP_K in TOP_K_VALUES:

        # ---------------------------------------------------------------
        # LASSO baseline
        # ---------------------------------------------------------------

        lasso_features_rep = (
            lasso_rank_rep
            .head(TOP_K)["Feature"]
            .tolist()
        )


        # ---------------------------------------------------------------
        # LASSO + DODA
        # ---------------------------------------------------------------

        doda_selector_rep = make_doda_lasso(
            TOP_K
        )

        doda_selector_rep.fit(
            X_train_rep_scaled,
            y_train_rep
        )

        doda_features_rep = list(
            doda_selector_rep.get_selected_features()
        )


        feature_sets = {
            "LASSO": lasso_features_rep,
            "LASSO + DODA": doda_features_rep
        }


        for method, selected_features in feature_sets.items():

            for model_name, model_template in cv_models.items():

                model = type(model_template)(
                    **model_template.get_params()
                )

                Xtr = X_train_rep_scaled[
                    selected_features
                ]

                Xva = X_val_rep_scaled[
                    selected_features
                ]


                model.fit(
                    Xtr,
                    y_train_rep
                )

                y_pred = model.predict(
                    Xva
                )

                y_prob = model.predict_proba(
                    Xva
                )[:, 1]


                repeated_cv_results.append({

                    "Run": run_id,
                    "Method": method,
                    "Top_K": TOP_K,
                    "Model": model_name,

                    "Accuracy": accuracy_score(
                        y_val_rep,
                        y_pred
                    ),

                    "Precision": precision_score(
                        y_val_rep,
                        y_pred,
                        zero_division=0
                    ),

                    "Recall": recall_score(
                        y_val_rep,
                        y_pred,
                        zero_division=0
                    ),

                    "F1": f1_score(
                        y_val_rep,
                        y_pred,
                        zero_division=0
                    ),

                    "ROC_AUC": roc_auc_score(
                        y_val_rep,
                        y_prob
                    ),

                    "Selected_Features": ", ".join(
                        selected_features
                    )
                })


repeated_cv_results_df = pd.DataFrame(
    repeated_cv_results
)

print("\nCompleted repeated CV.")
print(
    "Rows:",
    len(repeated_cv_results_df)
)

display(
    repeated_cv_results_df.head()
)


RUN 1/25
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x000001AC4C46AF00>
DODA Pipeline Started
Running: LogisticRegression
DODA feature order:
['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension']
Number of features: 30
Number of coefficients: 30
Operator Scores:
{'LogisticRegression_1': {'mean radius': np.float64(0.0), 'mean texture': np.float64(0.0), 'mean perimeter': np.float64(0.0), 'mean area': np.float64(0.0), 'mean smoothness': np.float64(0.0), 'mean compactness': np.float64(0.0), 'mean concavity': np.float64(0.0), 'mean concave points': np.float64(0.3919984806350876), 'mean symmetry': np.float64(0.0), 'mean fractal dimension': np.float64(0.0), 'radius error': np.float64(0.5389807424230852), 'texture error': np.float64(0.0), 'perimeter error': np.float64(0.0), 'area error': np.float64(0.0), 'smoot

,Run,Method,Top_K,Model,Accuracy,Precision,Recall,F1,ROC_AUC,Selected_Features
0,1,LASSO,5,Logistic Regression,0.947368,0.985075,0.929577,0.956522,0.995742,"worst radius, worst concave points, worst text..."
1,1,LASSO,5,Random Forest,0.956140,0.985294,0.943662,0.964029,0.997052,"worst radius, worst concave points, worst text..."
2,1,LASSO,5,XGBoost,0.947368,0.971014,0.943662,0.957143,0.995414,"worst radius, worst concave points, worst text..."
3,1,LASSO + DODA,5,Logistic Regression,0.947368,0.985075,0.929577,0.956522,0.995742,"worst radius, worst concave points, worst text..."
4,1,LASSO + DODA,5,Random Forest,0.956140,0.985294,0.943662,0.964029,0.996233,"worst radius, worst concave points, worst text..."


In [33]:
# =============================================================================
# STEP 25: REPEATED CV PERFORMANCE SUMMARY — MEAN ± STD
# =============================================================================

repeated_cv_summary_df = (
    repeated_cv_results_df
    .groupby(
        ["Method", "Top_K", "Model"]
    )
    .agg(
        Accuracy_Mean=("Accuracy", "mean"),
        Accuracy_STD=("Accuracy", "std"),

        Precision_Mean=("Precision", "mean"),
        Precision_STD=("Precision", "std"),

        Recall_Mean=("Recall", "mean"),
        Recall_STD=("Recall", "std"),

        F1_Mean=("F1", "mean"),
        F1_STD=("F1", "std"),

        ROC_AUC_Mean=("ROC_AUC", "mean"),
        ROC_AUC_STD=("ROC_AUC", "std")
    )
    .reset_index()
)

print("=" * 100)
print(
    "5×5 REPEATED STRATIFIED CV PERFORMANCE SUMMARY"
)
print("=" * 100)

display(
    repeated_cv_summary_df.round(4)
)

os.makedirs(
    "../../../results/breast_cancer/cv",
    exist_ok=True
)

repeated_cv_results_df.to_csv(
    "../../../results/breast_cancer/cv/"
    "lasso_doda_5x5_repeated_results.csv",
    index=False
)

repeated_cv_summary_df.to_csv(
    "../../../results/breast_cancer/cv/"
    "lasso_doda_5x5_repeated_summary.csv",
    index=False
)


5×5 REPEATED STRATIFIED CV PERFORMANCE SUMMARY


,Method,Top_K,Model,Accuracy_Mean,Accuracy_STD,Precision_Mean,Precision_STD,Recall_Mean,Recall_STD,F1_Mean,F1_STD,ROC_AUC_Mean,ROC_AUC_STD
0,LASSO,5,Logistic Regression,0.9592,0.0121,0.9701,0.0180,0.9653,0.0171,0.9675,0.0096,0.9930,0.0056
1,LASSO,5,Random Forest,0.9575,0.0133,0.9666,0.0221,0.9664,0.0229,0.9661,0.0106,0.9909,0.0061
2,LASSO,5,XGBoost,0.9550,0.0146,0.9679,0.0202,0.9608,0.0262,0.9639,0.0121,0.9918,0.0049
3,LASSO,10,Logistic Regression,0.9687,0.0139,0.9723,0.0147,0.9782,0.0134,0.9752,0.0110,0.9953,0.0037
4,LASSO,10,Random Forest,0.9564,0.0171,0.9721,0.0175,0.9585,0.0276,0.9649,0.0141,0.9925,0.0047
5,LASSO,10,XGBoost,0.9631,0.0143,0.9712,0.0139,0.9703,0.0259,0.9704,0.0120,0.9943,0.0035
6,LASSO,15,Logistic Regression,0.9698,0.0127,0.9750,0.0131,0.9770,0.0127,0.9759,0.0102,0.9951,0.0037
7,LASSO,15,Random Forest,0.9582,0.0181,0.9707,0.0207,0.9630,0.0279,0.9664,0.0147,0.9929,0.0044
8,LASSO,15,XGBoost,0.9631,0.0152,0.9718,0.0176,0.9697,0.0227,0.9705,0.0123,0.9942,0.0039
9,LASSO,20,Logistic Regression,0.9701,0.0132,0.9766,0.0141,0.9759,0.0149,0.9762,0.0105,0.9949,0.0043


## 4. Feature-Selection Stability and LASSO–DODA Agreement

The next analyses use the **25 repeated-CV runs** to examine how consistently features are selected.

We distinguish:

- **Selection frequency:** how often a feature is selected.
- **Jaccard similarity:** overlap between feature sets.
- **Kuncheva index:** chance-adjusted agreement for fixed-size feature subsets.
- **LASSO–DODA agreement:** whether DODA changes the selected set relative to the LASSO statistical baseline.

These analyses evaluate **feature-selection stability and set membership**, not merely predictive performance.


In [34]:
# =============================================================================
# STEP 26: PREPARE STABILITY DATAFRAME
# =============================================================================

stability_df = repeated_cv_results_df[
    [
        "Run",
        "Method",
        "Top_K",
        "Selected_Features"
    ]
].copy()

stability_df["Feature_Set"] = (
    stability_df["Selected_Features"]
    .apply(
        lambda x: {
            feature.strip()
            for feature in x.split(",")
            if feature.strip()
        }
    )
)

feature_names = X.columns.tolist()
N_FEATURES = len(feature_names)

print("=" * 80)
print("STABILITY ANALYSIS PREPARATION")
print("=" * 80)

print("Rows:", stability_df.shape[0])
print(
    "Methods:",
    stability_df["Method"].unique()
)
print(
    "Top-K:",
    sorted(stability_df["Top_K"].unique())
)
print(
    "Total predictors:",
    N_FEATURES
)

display(
    stability_df[
        [
            "Run",
            "Method",
            "Top_K",
            "Feature_Set"
        ]
    ].head()
)


STABILITY ANALYSIS PREPARATION
Rows: 600
Methods: <StringArray>
['LASSO', 'LASSO + DODA']
Length: 2, dtype: str
Top-K: [np.int64(5), np.int64(10), np.int64(15), np.int64(20)]
Total predictors: 30


,Run,Method,Top_K,Feature_Set
0,1,LASSO,5,"{worst concave points, radius error, mean conc..."
1,1,LASSO,5,"{worst concave points, radius error, mean conc..."
2,1,LASSO,5,"{worst concave points, radius error, mean conc..."
3,1,LASSO + DODA,5,"{worst concave points, radius error, mean conc..."
4,1,LASSO + DODA,5,"{worst concave points, radius error, mean conc..."


In [35]:
# =============================================================================
# STEP 27: TEST 1 — FEATURE SELECTION FREQUENCY
# =============================================================================

selection_frequency = []


for method in sorted(
    stability_df["Method"].unique()
):

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ]

        total_runs = len(subset)

        all_features = sorted(
            set().union(
                *subset["Feature_Set"]
            )
        )

        for feature in all_features:

            selected_count = sum(
                feature in feature_set
                for feature_set in subset["Feature_Set"]
            )

            selection_frequency.append({

                "Method": method,

                "Top_K": top_k,

                "Feature": feature,

                "Selected_Count": selected_count,

                "Total_Runs": total_runs,

                "Selection_Frequency_Percentage":
                    (
                        selected_count /
                        total_runs
                    ) * 100
            })


selection_frequency_df = pd.DataFrame(
    selection_frequency
)

selection_frequency_df = (
    selection_frequency_df
    .sort_values(
        by=[
            "Method",
            "Top_K",
            "Selection_Frequency_Percentage"
        ],
        ascending=[
            True,
            True,
            False
        ]
    )
    .reset_index(drop=True)
)

display(
    selection_frequency_df
)


,Method,Top_K,Feature,Selected_Count,Total_Runs,Selection_Frequency_Percentage
0,LASSO,5,worst concave points,75,75,100.0
1,LASSO,5,worst radius,75,75,100.0
2,LASSO,5,worst texture,75,75,100.0
3,LASSO,5,mean concave points,69,75,92.0
4,LASSO,5,radius error,60,75,80.0
...,...,...,...,...,...,...
119,LASSO + DODA,20,area error,69,75,92.0
120,LASSO + DODA,20,worst symmetry,69,75,92.0
121,LASSO + DODA,20,smoothness error,39,75,52.0
122,LASSO + DODA,20,worst concavity,33,75,44.0


In [36]:
# =============================================================================
# STEP 28: TOP FREQUENTLY SELECTED FEATURES
# =============================================================================

top_frequency_features = (
    selection_frequency_df
    .groupby(
        ["Method", "Top_K"],
        group_keys=False
    )
    .head(10)
    .reset_index(drop=True)
)

print("=" * 80)
print("TOP FREQUENTLY SELECTED FEATURES")
print("=" * 80)

display(
    top_frequency_features
)


TOP FREQUENTLY SELECTED FEATURES


,Method,Top_K,Feature,Selected_Count,Total_Runs,Selection_Frequency_Percentage
0,LASSO,5,worst concave points,75,75,100.0
1,LASSO,5,worst radius,75,75,100.0
2,LASSO,5,worst texture,75,75,100.0
3,LASSO,5,mean concave points,69,75,92.0
4,LASSO,5,radius error,60,75,80.0
...,...,...,...,...,...,...
72,LASSO + DODA,20,mean perimeter,75,75,100.0
73,LASSO + DODA,20,mean radius,75,75,100.0
74,LASSO + DODA,20,mean smoothness,75,75,100.0
75,LASSO + DODA,20,mean symmetry,75,75,100.0


In [37]:
# =============================================================================
# STEP 29: TEST 2 — PAIRWISE JACCARD SIMILARITY
# =============================================================================

jaccard_results = []


for method in sorted(
    stability_df["Method"].unique()
):

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ].reset_index(drop=True)

        feature_sets = (
            subset["Feature_Set"]
            .tolist()
        )

        pairwise_scores = []


        for set_a, set_b in combinations(
            feature_sets,
            2
        ):

            intersection_size = len(
                set_a.intersection(set_b)
            )

            union_size = len(
                set_a.union(set_b)
            )

            jaccard_score = (
                intersection_size /
                union_size
                if union_size
                else 1.0
            )

            pairwise_scores.append(
                jaccard_score
            )


        if pairwise_scores:

            jaccard_results.append({

                "Method": method,

                "Top_K": top_k,

                "Mean_Jaccard": np.mean(
                    pairwise_scores
                ),

                "Std_Jaccard": np.std(
                    pairwise_scores,
                    ddof=1
                ),

                "Minimum_Jaccard": np.min(
                    pairwise_scores
                ),

                "Maximum_Jaccard": np.max(
                    pairwise_scores
                )
            })


jaccard_df = pd.DataFrame(
    jaccard_results
)

print("=" * 80)
print("PAIRWISE JACCARD SIMILARITY")
print("=" * 80)

display(
    jaccard_df.round(4)
)


PAIRWISE JACCARD SIMILARITY


,Method,Top_K,Mean_Jaccard,Std_Jaccard,Minimum_Jaccard,Maximum_Jaccard
0,LASSO,5,0.8386,0.1744,0.4286,1.0
1,LASSO,10,0.8076,0.1233,0.5385,1.0
2,LASSO,15,0.8704,0.0902,0.6667,1.0
3,LASSO,20,0.9004,0.0698,0.7391,1.0
4,LASSO + DODA,5,0.8443,0.1663,0.6667,1.0
5,LASSO + DODA,10,0.8428,0.1098,0.6667,1.0
6,LASSO + DODA,15,0.8960,0.0784,0.7647,1.0
7,LASSO + DODA,20,0.9237,0.0608,0.8182,1.0


In [38]:
# =============================================================================
# STEP 30: TEST 3 — KUNCHEVA INDEX
# =============================================================================

kuncheva_results = []


for method in sorted(
    stability_df["Method"].unique()
):

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ].reset_index(drop=True)

        feature_sets = (
            subset["Feature_Set"]
            .tolist()
        )


        for i in range(
            len(feature_sets)
        ):

            for j in range(
                i + 1,
                len(feature_sets)
            ):

                set_a = feature_sets[i]
                set_b = feature_sets[j]

                intersection = len(
                    set_a.intersection(set_b)
                )

                denominator = (
                    top_k *
                    (
                        N_FEATURES -
                        top_k
                    )
                )


                if denominator == 0:

                    kuncheva_value = np.nan

                else:

                    kuncheva_value = (
                        (
                            intersection *
                            N_FEATURES
                        )
                        -
                        (
                            top_k ** 2
                        )
                    ) / denominator


                kuncheva_results.append({

                    "Method": method,

                    "Top_K": top_k,

                    "Run_A": subset.loc[
                        i,
                        "Run"
                    ],

                    "Run_B": subset.loc[
                        j,
                        "Run"
                    ],

                    "Intersection": intersection,

                    "Kuncheva_Index": kuncheva_value
                })


kuncheva_df = pd.DataFrame(
    kuncheva_results
)

kuncheva_summary = (
    kuncheva_df
    .groupby(
        ["Method", "Top_K"]
    )["Kuncheva_Index"]
    .agg(
        Mean_Kuncheva="mean",
        Std_Kuncheva="std",
        Min_Kuncheva="min",
        Max_Kuncheva="max"
    )
    .reset_index()
)


print("=" * 80)
print("KUNCHEVA INDEX SUMMARY")
print("=" * 80)

display(
    kuncheva_summary.round(4)
)


KUNCHEVA INDEX SUMMARY


,Method,Top_K,Mean_Kuncheva,Std_Kuncheva,Min_Kuncheva,Max_Kuncheva
0,LASSO,5,0.8825,0.1290,0.5200,1.0
1,LASSO,10,0.8326,0.1128,0.5500,1.0
2,LASSO,15,0.8564,0.1032,0.6000,1.0
3,LASSO,20,0.8385,0.1161,0.5500,1.0
4,LASSO + DODA,5,0.8879,0.1198,0.7600,1.0
5,LASSO + DODA,10,0.8662,0.0973,0.7000,1.0
6,LASSO + DODA,15,0.8867,0.0877,0.7333,1.0
7,LASSO + DODA,20,0.8779,0.0992,0.7000,1.0


In [39]:
# =============================================================================
# STEP 31: TEST 4 — LASSO vs DODA FEATURE-SET AGREEMENT
# =============================================================================
# This compares feature-set membership within the SAME repeated-CV run
# and Top-K value.

agreement_results = []


for run in sorted(
    stability_df["Run"].unique()
):

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        lasso_rows = stability_df[
            (stability_df["Run"] == run) &
            (stability_df["Method"] == "LASSO") &
            (stability_df["Top_K"] == top_k)
        ]

        doda_rows = stability_df[
            (stability_df["Run"] == run) &
            (stability_df["Method"] == "LASSO + DODA") &
            (stability_df["Top_K"] == top_k)
        ]


        if (
            lasso_rows.empty or
            doda_rows.empty
        ):
            continue


        lasso_set = (
            lasso_rows.iloc[0]["Feature_Set"]
        )

        doda_set = (
            doda_rows.iloc[0]["Feature_Set"]
        )


        intersection_size = len(
            lasso_set.intersection(
                doda_set
            )
        )

        union_size = len(
            lasso_set.union(
                doda_set
            )
        )

        jaccard_agreement = (
            intersection_size /
            union_size
            if union_size
            else 1.0
        )


        agreement_results.append({

            "Run": run,

            "Top_K": top_k,

            "LASSO_Features":
                ", ".join(
                    sorted(lasso_set)
                ),

            "DODA_Features":
                ", ".join(
                    sorted(doda_set)
                ),

            "Common_Feature_Count":
                intersection_size,

            "LASSO_Feature_Count":
                len(lasso_set),

            "DODA_Feature_Count":
                len(doda_set),

            "Same_Feature_Set":
                lasso_set == doda_set,

            "Jaccard_Agreement":
                jaccard_agreement,

            "Features_Added_by_DODA":
                ", ".join(
                    sorted(
                        doda_set -
                        lasso_set
                    )
                ),

            "Features_Removed_by_DODA":
                ", ".join(
                    sorted(
                        lasso_set -
                        doda_set
                    )
                )
        })


agreement_df = pd.DataFrame(
    agreement_results
)


print("=" * 100)
print("LASSO vs DODA FEATURE-SET AGREEMENT")
print("=" * 100)

display(
    agreement_df
)


LASSO vs DODA FEATURE-SET AGREEMENT


,Run,Top_K,LASSO_Features,DODA_Features,Common_Feature_Count,LASSO_Feature_Count,DODA_Feature_Count,Same_Feature_Set,Jaccard_Agreement,Features_Added_by_DODA,Features_Removed_by_DODA
0,1,5,"mean concave points, radius error, worst conca...","mean concave points, radius error, worst conca...",5,5,5,True,1.000000,,
1,1,10,"area error, compactness error, concave points ...","mean concave points, mean perimeter, mean radi...",7,10,10,False,0.538462,"mean perimeter, mean radius, mean texture","area error, compactness error, concave points ..."
2,1,15,"area error, compactness error, concave points ...","mean area, mean compactness, mean concave poin...",10,15,15,False,0.500000,"mean perimeter, mean radius, mean smoothness, ...","area error, compactness error, concave points ..."
3,1,20,"area error, compactness error, concave points ...","area error, mean area, mean compactness, mean ...",16,20,20,False,0.666667,"mean texture, perimeter error, smoothness erro...","compactness error, concave points error, conca..."
4,2,5,"mean concave points, worst concave points, wor...","mean concave points, worst concave points, wor...",5,5,5,True,1.000000,,
...,...,...,...,...,...,...,...,...,...,...,...
95,24,20,"area error, compactness error, concave points ...","area error, mean area, mean compactness, mean ...",16,20,20,False,0.666667,"mean texture, perimeter error, smoothness erro...","compactness error, concave points error, conca..."
96,25,5,"mean concave points, radius error, worst conca...","mean concave points, radius error, worst conca...",5,5,5,True,1.000000,,
97,25,10,"area error, compactness error, concave points ...","mean concave points, mean perimeter, mean radi...",7,10,10,False,0.538462,"mean perimeter, mean radius, mean texture","area error, compactness error, concave points ..."
98,25,15,"area error, compactness error, concave points ...","mean area, mean compactness, mean concave poin...",10,15,15,False,0.500000,"mean perimeter, mean radius, mean smoothness, ...","area error, compactness error, concave points ..."


In [40]:
# =============================================================================
# STEP 32: SUMMARY OF LASSO vs DODA AGREEMENT
# =============================================================================

agreement_summary_df = (
    agreement_df
    .groupby("Top_K")
    .agg(
        Same_Feature_Set_Percentage=(
            "Same_Feature_Set",
            lambda x: x.mean() * 100
        ),

        Mean_Jaccard_Agreement=(
            "Jaccard_Agreement",
            "mean"
        ),

        Std_Jaccard_Agreement=(
            "Jaccard_Agreement",
            "std"
        ),

        Mean_Common_Features=(
            "Common_Feature_Count",
            "mean"
        )
    )
    .reset_index()
)


print("=" * 100)
print(
    "FINAL LASSO vs DODA FEATURE-SET AGREEMENT SUMMARY"
)
print("=" * 100)

display(
    agreement_summary_df.round(4)
)


# -------------------------------------------------------------------------
# Save stability outputs
# -------------------------------------------------------------------------

os.makedirs(
    "../../../results/breast_cancer/stability",
    exist_ok=True
)

selection_frequency_df.to_csv(
    "../../../results/breast_cancer/stability/"
    "feature_selection_frequency_lasso.csv",
    index=False
)

jaccard_df.to_csv(
    "../../../results/breast_cancer/stability/"
    "jaccard_similarity_lasso.csv",
    index=False
)

kuncheva_df.to_csv(
    "../../../results/breast_cancer/stability/"
    "kuncheva_pairwise_lasso.csv",
    index=False
)

kuncheva_summary.to_csv(
    "../../../results/breast_cancer/stability/"
    "kuncheva_summary_lasso.csv",
    index=False
)

agreement_df.to_csv(
    "../../../results/breast_cancer/stability/"
    "lasso_doda_agreement.csv",
    index=False
)

agreement_summary_df.to_csv(
    "../../../results/breast_cancer/stability/"
    "lasso_doda_agreement_summary.csv",
    index=False
)


FINAL LASSO vs DODA FEATURE-SET AGREEMENT SUMMARY


,Top_K,Same_Feature_Set_Percentage,Mean_Jaccard_Agreement,Std_Jaccard_Agreement,Mean_Common_Features
0,5,84.0,0.9467,0.1247,4.84
1,10,0.0,0.6354,0.1095,7.72
2,15,0.0,0.5000,0.0000,10.00
3,20,0.0,0.6613,0.0185,15.92


In [42]:
# =============================================================================
# STEP 33: FINAL NOTEBOOK CONSISTENCY CHECK
# =============================================================================

print("=" * 80)
print("BREAST CANCER LASSO + DODA NOTEBOOK SANITY CHECK")
print("=" * 80)

print(
    "Dataset shape             :",
    df.shape
)

print(
    "Predictor count           :",
    X.shape[1]
)

print(
    "Target column             :",
    target_column
)

print(
    "Missing values in X       :",
    int(X.isnull().sum().sum())
)

print(
    "Fixed-split Top-K values  :",
    k_values
)

print(
    "CV Top-K values           :",
    TOP_K_VALUES
)

print(
    "Repeated-CV runs          :",
    N_SPLITS * N_REPEATS
)

print(
    "Baseline result rows      :",
    len(lasso_evaluation)
)

print(
    "DODA result rows          :",
    len(lasso_doda_results_df)
)

print(
    "5-fold CV result rows     :",
    len(cv_results_df)
)

print(
    "5×5 CV result rows        :",
    len(repeated_cv_results_df)
)

print(
    "Stability methods         :",
    stability_df["Method"].unique()
)


assert target_column == "target"
assert X.isnull().sum().sum() == 0
assert TOP_K_VALUES == [5, 10, 15, 20]
assert all(
    k <= X.shape[1]
    for k in TOP_K_VALUES
)
assert set(
    stability_df["Method"].unique()
) == {
    "LASSO",
    "LASSO + DODA"
}



BREAST CANCER LASSO + DODA NOTEBOOK SANITY CHECK
Dataset shape             : (569, 31)
Predictor count           : 30
Target column             : target
Missing values in X       : 0
Fixed-split Top-K values  : [5, 10, 15, 20]
CV Top-K values           : [5, 10, 15, 20]
Repeated-CV runs          : 25
Baseline result rows      : 12
DODA result rows          : 12
5-fold CV result rows     : 120
5×5 CV result rows        : 600
Stability methods         : <StringArray>
['LASSO', 'LASSO + DODA']
Length: 2, dtype: str


# 3. LASSO + DODA: Model Evaluation & Interpretation

---

## 3.1 Overall Predictive Performance

### 📊 Performance by Top-K

#### Top-K = 5

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:------|
| Logistic Regression | 0.9561 → 0.9561 | 0.9718 → 0.9718 | 0.9583 → 0.9583 | 0.9650 → 0.9650 | 0.9934 → 0.9934 | 🟡 Maintained |
| Random Forest | 0.9561 → 0.9561 | 0.9718 → 0.9589 | 0.9583 → 0.9722 | 0.9650 → 0.9655 | 0.9950 → 0.9950 | 🟡 Comparable |
| XGBoost | 0.9474 → **0.9649** | 0.9583 → **0.9722** | 0.9583 → **0.9722** | 0.9583 → **0.9722** | 0.9940 → 0.9937 | 🟢 Improved |

#### Top-K = 10

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:------|
| Logistic Regression | 0.9737 → 0.9386 | 0.9859 → 0.9851 | 0.9722 → 0.9167 | 0.9790 → 0.9496 | 0.9960 → 0.9947 | 🔴 Reduced |
| Random Forest | 0.9649 → 0.9386 | 0.9857 → 0.9710 | 0.9583 → 0.9306 | 0.9718 → 0.9504 | 0.9947 → 0.9947 | 🔴 Reduced |
| XGBoost | 0.9561 → 0.9561 | 0.9589 → 0.9718 | 0.9722 → 0.9583 | 0.9655 → 0.9650 | 0.9924 → 0.9947 | 🟡 Comparable |

#### Top-K = 15

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:------|
| Logistic Regression | 0.9474 → **0.9561** | 0.9853 → 0.9855 | 0.9306 → **0.9444** | 0.9571 → **0.9645** | 0.9937 → 0.9937 | 🟢 Improved |
| Random Forest | 0.9474 → **0.9561** | 0.9714 → 0.9718 | 0.9444 → **0.9583** | 0.9577 → **0.9650** | 0.9947 → 0.9942 | 🟢 Improved |
| XGBoost | 0.9649 → 0.9561 | 0.9722 → 0.9718 | 0.9722 → 0.9583 | 0.9722 → 0.9650 | 0.9950 → 0.9944 | 🔴 Reduced |

#### Top-K = 20

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:------|
| Logistic Regression | 0.9561 → 0.9474 | 0.9855 → 0.9853 | 0.9444 → 0.9306 | 0.9645 → 0.9571 | 0.9947 → 0.9947 | 🔴 Reduced |
| Random Forest | 0.9561 → 0.9474 | 0.9718 → 0.9583 | 0.9583 → 0.9583 | 0.9650 → 0.9583 | 0.9940 → 0.9937 | 🔴 Reduced |
| XGBoost | 0.9737 → 0.9474 | 0.9859 → 0.9459 | 0.9722 → 0.9722 | 0.9790 → 0.9589 | 0.9967 → 0.9964 | 🔴 Reduced |

> **Notation:** `LASSO → DODA`  
> 🟢 Improved &nbsp;&nbsp; 🟡 Maintained / comparable &nbsp;&nbsp; 🔴 Reduced

---

### 🔎 Overall Performance Pattern

| DODA Effect | Configurations |
|:------------|:---------------|
| 🟢 **Improved** | XGBoost Top-K 5; Logistic Regression & Random Forest Top-K 15 |
| 🟡 **Maintained / Comparable** | Logistic Regression & Random Forest Top-K 5; XGBoost Top-K 10 |
| 🔴 **Reduced** | Logistic Regression & Random Forest Top-K 10; XGBoost Top-K 15; all models Top-K 20 |

**Key finding:** DODA does not consistently improve predictive performance. Its effect depends on the model and Top-K feature size.

The strongest LASSO baseline results were:

- **Logistic Regression, Top-K = 10:** Accuracy = **0.9737**, F1 = **0.9790**
- **XGBoost, Top-K = 20:** Accuracy = **0.9737**, F1 = **0.9790**, ROC-AUC = **0.9967**

The strongest DODA result by accuracy and F1-score was:

- **XGBoost, Top-K = 5:** Accuracy = **0.9649**, F1 = **0.9722**

---

## 3.2 Effect of DODA on Feature Selection

### 📊 Top-10 Ranking Comparison

| Rank | LASSO | DODA | Change |
|-----:|:------|:-----|:------|
| 1 | `worst radius` | `worst radius` | — |
| 2 | `worst concave points` | `worst concave points` | — |
| 3 | `worst texture` | `worst texture` | — |
| 4 | `radius error` | `mean concave points` | ⬆️ |
| 5 | `mean concave points` | `radius error` | ⬇️ |
| 6 | `worst symmetry` | `worst symmetry` | — |
| 7 | `worst smoothness` | `worst smoothness` | — |
| 8 | `mean texture` | `mean texture` | — |
| 9 | `mean symmetry` | `mean radius` | ↕️ |
| 10 | `mean smoothness` | `mean perimeter` | ↕️ |

### 🎯 Main Feature Changes

- **Top-5 feature set:** unchanged, but internal ranking changed.
- `mean concave points`: **Rank 5 → 4** ⬆️
- `radius error`: **Rank 4 → 5** ⬇️
- `mean radius` and `mean perimeter` entered the DODA Top-10.
- `mean concavity` entered the DODA Top-15.
- `smoothness error` entered the DODA Top-20.
- Several later changes occurred among features with zero mathematical and DODA scores and should therefore be interpreted cautiously.

---

## 3.3 Clinical Weight Alignment

| Feature | LASSO Rank | DODA Rank | Clinical Weight | Direction |
|:--------|-----------:|----------:|----------------:|:---------:|
| `mean concave points` | 5 | **4** | **0.9** | ⬆️ |
| `radius error` | 4 | 5 | 0.5 | ⬇️ |
| `mean radius` | 14 | **9** | **0.9** | ⬆️ |
| `mean perimeter` | 13 | **10** | **0.9** | ⬆️ |
| `mean concavity` | 19 | **14** | **0.9** | ⬆️ |
| `mean symmetry` | 9 | 15 | 0.5 | ⬇️ |
| `mean smoothness` | 10 | 12 | 0.4 | ⬇️ |

### 🩺 Interpretation

The clearest clinically aligned scored change was:

```text
radius error          → Rank 4 → 5
mean concave points   → Rank 5 → 4